controllers/effective_dnspolicies_reconciler.go

```diff
@@ -0,0 +1,268 @@
+				lLogger.Error(err, "unable to destruct dns record") // should never happen
+				continue
+			}
+			resource := r.client.Resource(DNSRecordResource).Namespace(desiredRecord.GetNamespace())
+
+			existingRecordObj, ok := lo.Find(topology.Objects().Children(listener), func(o machinery.Object) bool {
+				_, ok := o.(*controller.RuntimeObject).Object.(*kuadrantdnsv1alpha1.DNSRecord)
+				return ok && o.GetNamespace() == listener.GetNamespace() && o.GetName() == dnsRecordName(listener.Gateway.Name, string(listener.Name))
+			})
+
+			//Update
+			if ok {
+				existingRecord := existingRecordObj.(*controller.RuntimeObject).Object.(*kuadrantdnsv1alpha1.DNSRecord)
+
+				//Deal with the potential deletion of a record first
+				if !hasAttachedRoute || len(desiredRecord.Spec.Endpoints) == 0 {
+					if logger.V(1).Enabled() {
+						if !hasAttachedRoute {
+							lLogger.V(1).Info("listener has no attached routes, deleting DNS record for listener")
+						} else {
+							lLogger.V(1).Info("no endpoint addresses for DNSRecord, deleting DNS record for listener")
+						}
+					}
+					r.deleteRecord(ctx, existingRecordObj)
+					continue
+				}
+
+				if desiredRecord.Spec.RootHost != existingRecord.Spec.RootHost {
+					lLogger.V(1).Info("listener hostname has changed, deleting DNS record for listener")
+					r.deleteRecord(ctx, existingRecordObj)
+					//Break to allow it to try the creation of the desired record
+					break
+				}
+
+				if reflect.DeepEqual(existingRecord.Spec, desiredRecord.Spec) &&
+					reflect.DeepEqual(existingRecord.OwnerReferences, desiredRecord.OwnerReferences) {
+					lLogger.V(1).Info("dns record is up to date, nothing to do")
+					continue
+				}
+
+				lLogger.V(1).Info("updating DNS record for listener")
+				_, err = resource.Update(ctx, obj, metav1.UpdateOptions{})
+				if err != nil {
+					lLogger.Error(err, "unable to update dns record")
+				}
+				continue
+			}
+
+			if !hasAttachedRoute || len(desiredRecord.Spec.Endpoints) == 0 {
+				continue
+			}
+
+			//Create
+			lLogger.V(1).Info("creating DNS record for listener")
+			_, err = resource.Create(ctx, obj, metav1.CreateOptions{})
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				lLogger.Error(err, "unable to create dns record")
+			}
+		}
+
+		if !gatewayHasAttachedRoutes {
+			pLogger.V(1).Info("gateway has no attached routes")
+			//return ErrNoRoutes
+		}
+	}
+
+	return r.deleteOrphanDNSRecords(controller.LoggerIntoContext(ctx, logger), topology)
+}
+
+// listenersForPolicy returns an array of listeners that are targeted by the given policy.
+// If the target is a Listener a single element array containing that listener is returned.
+// If the target is a Gateway all listeners that do not have a DNS policy explicitly attached are returned.
+func (r *EffectiveDNSPoliciesReconciler) listenersForPolicy(_ context.Context, topology *machinery.Topology, policy machinery.Policy, policyTypeFilterFunc dnsPolicyTypeFilter) []*machinery.Listener {
```

Comment:
so with this do we get section name support then?

Reply:
Yes, this should allow section name to be added without new changes to this task. I have already got this working and will create a separate PR for that after this one is merged.